OptimizelySDK.Tests/DecisionServiceTest.cs

```diff
@@ -0,0 +1,341 @@
+            {
+                new TrafficAllocation {EntityId = "1", EndOfRange = 1000 }
+            };
+
+            var userIdToVariationKeyMap = new Dictionary<string, string>
+            {
+                {userId, invalidVariationKey }
+            };
+
+            var experiment = new Experiment
+            {
+                Id = "1234",
+                Key = "exp_key",
+                Status = "Running",
+                LayerId = "1",
+                AudienceIds = new string[0],
+                Variations = variations,
+                TrafficAllocation = trafficAllocation,
+                ForcedVariations = userIdToVariationKeyMap
+            };
+
+            //Experiment experiment = new Experiment("1234", "exp_key", "Running", "1", Collections.< String > emptyList(),
+            //        variations, userIdToVariationKeyMap, trafficAllocations);
+            Assert.IsNull(decisionService.GetWhitelistedVariation(experiment, userId));
+            LoggerMock.Verify(l => l.Log(LogLevel.ERROR,
+                string.Format("Variation \"{0}\" is not in the datafile. Not activating user \"{1}\".", invalidVariationKey, userId)), 
+                Times.Once);
+        }
+
+        [Test]
+        public void GetForcedVariationReturnsNullWhenUserIsNotWhitelisted()
+        {
+            Bucketer bucketer = new Bucketer(LoggerMock.Object);
+            DecisionService decisionService = new DecisionService(bucketer, ErrorHandlerMock.Object, ValidProjectConfig, null, LoggerMock.Object);
+
+            Assert.IsNull(decisionService.GetWhitelistedVariation(WhitelistedExperiment, GenericUserId));
+        }
+
+        [Test]
+        public void BucketReturnsVariationStoredInUserProfile()
+        {
+            Experiment experiment = NoAudienceProjectConfig.Experiments[0];
+            Variation variation = experiment.Variations[0];
+            Decision decision = new Decision(variation.Id);
+
+            UserProfile userProfile = new UserProfile(UserProfileId, new Dictionary<string, Decision>
+            {
+                { experiment.Id, decision }
+            });
+
+            UserProfileServiceMock.Setup(_ => _.Lookup(UserProfileId)).Returns(userProfile.ToMap());
+
+            Bucketer bucketer = new Bucketer(LoggerMock.Object);
+            DecisionService decisionService = new DecisionService(bucketer, ErrorHandlerMock.Object, NoAudienceProjectConfig, UserProfileServiceMock.Object, LoggerMock.Object);
+
+            Assert.IsTrue(TestData.CompareObjects(variation, decisionService.GetVariation(experiment, UserProfileId, new UserAttributes())));
+
+            LoggerMock.Verify(l => l.Log(LogLevel.INFO, string.Format("Returning previously activated variation \"{0}\" of experiment \"{1}\" for user \"{2}\" from user profile.",
+                variation.Key, experiment.Key, UserProfileId)));
+
+
+
+
+            //    UserProfileService userProfileService = mock(UserProfileService.class);
```

Comment:
Mike this test method returns the variation which is already stored. BucketReturnsVariationStoredInUserProfile. This portion is returning the userprofile. 
userProfile = UserProfileUtil.ConvertMapToUserProfile(userProfileMap);
                        variation = GetStoredVariation(experiment, userProfile);
                        if (variation != null) return variation;